### PR TITLE
[Off-chain] Simplify `TxClient` with `EventsQueryClient`

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Human Summary

- Simplifies `TxClient` implementation by replacing `EventsQueryClient` usage with an `EventsReplayClient`.
- Simplifies `EventsReplayClient` to take a single generic type parameter.

## Issue

- #256 

## Type of change

Select one or more:

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Run E2E tests locally**: `make test_e2e`
- [ ] **Run E2E tests on DevNet**: Add the `devnet-test-e2e` label to the PR. This is VERY expensive, only do it after all the reviews are complete.

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

**Comment by bryanchriswhite** [comment]
I'm not sure why the E2E tests are failing in CI. :thinking: 

(cc  @okdas, if you have bandwidth, I would be grateful for an explanation at least :sweat_smile: :shrug:)

**Comment by okdas** [comment]
> I'm not sure why the E2E tests are failing in CI. 🤔
> 
> (cc @okdas, if you have bandwidth, I would be grateful for an explanation at least 😅 🤷)

Interesting! RelayMiner is down with the following error:

```
{"level":"info","message":"Starting relay miner..."}
{"level":"info","message":"starting relayer sessions manager"}
{"level":"info","message":"starting relayer proxy"}
{"level":"info","proxy host":"0.0.0.0:8545","message":"starting relay proxy server"}
{"level":"info","message":"starting ring cache"}
panic: unexpected error deserialising event: {"jsonrpc":"2.0","result":{},"id":"XU72jMrpfPM="}
: attempted to deserialize non-tx event bytes [/go/pkg/mod/cosmossdk.io/errors@v1.0.0-beta.7/errors.go:153]; eventBz: {"jsonrpc":"2.0","result":{},"id":"XU72jMrpfPM="}


goroutine 208 [running]:
github.com/pokt-network/poktroll/pkg/client/events.(*replayClient[...]).newMapEventsBytesToTFn.func1({{0xc00038b800?, 0x32?, 0x200}, {0x0, 0x0}})
    /poktroll/pkg/client/events/replay_client.go:295 +0x1b4
github.com/pokt-network/poktroll/pkg/observable/channel.goMapTransformNotification[...]({0x40c4408?, 0xc000a09590}, {0x40bc750, 0xc000ff0ea0?}, 0xc001234460?, 0xc00120e618)
    /poktroll/pkg/observable/channel/map.go:117 +0xff
created by github.com/pokt-network/poktroll/pkg/observable/channel.MapReplay[...]
    /poktroll/pkg/observable/channel/map.go:76 +0x1da
Stream closed EOF for devnet-issue-330/devnet-issue-330-relayminer-59fc47499f-l54xl (relayminer)
```

Does that look familiar? I'll see how to better provide this kind of information to you and other developers!

**Review by h5law — APPROVED** [review]
LGTM Love the simplification <3 lets get this merged

File: pkg/client/tx/client.go [hunk 1]
```diff
@@ -326,69 +336,30 @@ func (tClient *txClient) addPendingTransactions(
 	return either.AsyncErr(errCh)
 }
 
-// subscribeToOwnTxs establishes an event query subscription to monitor transactions
-// originating from this client's signing address.
+// goSubscribeToOwnTxs establishes an event query subscription to monitor transactions
```
Comment (Olshansk) [inline pkg/client/tx/client.go:339]: This is 🔥

**Review by Olshansk — APPROVED** [review]
@bryanchriswhite Great PR (small, well scoped, clean, etc...).

I left some comments around the choice of wording of `event` vs `result`, but after sleeping on it, I think I could argue both ways. Feel free to accept or reject the changes before merging in.

File: pkg/client/tx/client.go [hunk 2]
```diff
@@ -571,3 +483,24 @@ func (tClient *txClient) getTxTimeoutError(ctx context.Context, txHashHex string
...
+
+// unmarshalTxResult attempts to deserialize a slice of bytes into a TxEvent.
+// It checks if the given bytes correspond to a valid transaction event.
+// If the resulting TxEvent has empty transaction bytes, it assumes that
+// the message was not a transaction event and returns an ErrNonTxEventBytes error.
+func unmarshalTxResult(eventBz []byte) (*abci.TxResult, error) {
+	txResult := new(abci.TxResult)
+
+	// Try to deserialize the provided bytes into a TxEvent.
+	if err := json.Unmarshal(eventBz, txResult); err != nil {
```
Comment (bryanchriswhite) [inline pkg/client/tx/client.go:495]:
```suggestion
	if err := json.Unmarshal(txResultBz, txResult); err != nil {
```

File: pkg/client/tx/client.go [hunk 3]
```diff
@@ -326,69 +336,30 @@ func (tClient *txClient) addPendingTransactions(
...
-	for eitherTxEvent := range txEventsObserver.Ch() {
-		txEvent, err := eitherTxEvent.ValueOrError()
-		if err != nil {
-			return
-		}
-
+func (tClient *txClient) goSubscribeToOwnTxs(ctx context.Context) {
+	txEventsObs := tClient.eventsReplayClient.EventsSequence(ctx)
+	txEventsCh := txEventsObs.Subscribe(ctx).Ch()
+	for txResult := range txEventsCh {
```
Comment (bryanchriswhite) [inline pkg/client/tx/client.go:360]:
```suggestion
	txResultsObs := tClient.eventsReplayClient.EventsSequence(ctx)
	txResultsCh := txResultsObs.Subscribe(ctx).Ch()
	for txResult := range txResultsCh {
```

File: pkg/client/tx/client.go [hunk 4]
```diff
@@ -571,3 +483,24 @@ func (tClient *txClient) getTxTimeoutError(ctx context.Context, txHashHex string
...
+// the message was not a transaction event and returns an ErrNonTxEventBytes error.
+func unmarshalTxResult(eventBz []byte) (*abci.TxResult, error) {
+	txResult := new(abci.TxResult)
+
+	// Try to deserialize the provided bytes into a TxEvent.
+	if err := json.Unmarshal(eventBz, txResult); err != nil {
+		return nil, events.ErrEventsUnmarshalEvent.Wrap(err.Error())
+	}
+
+	// Check if the TxEvent has empty transaction bytes, which indicates
```
Comment (bryanchriswhite) [inline pkg/client/tx/client.go:499]:
```suggestion
	// Check if the TxResult has empty transaction bytes, which indicates
```

File: pkg/client/tx/client_test.go [hunk 5]
```diff
@@ -101,12 +106,14 @@ func TestTxClient_SignAndBroadcast_Succeeds(t *testing.T) {
 	require.NoError(t, err)
 
 	// Construct the expected transaction event bytes from the expected transaction bytes.
-	txEventBz, err := json.Marshal(&tx.TxEvent{Tx: expectedTx})
+	txEventBz, err := json.Marshal(abci.TxResult{Tx: expectedTx})
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:109]:
```suggestion
	txResultBz, err := json.Marshal(abci.TxResult{Tx: expectedTx})
```

File: pkg/client/tx/client_test.go [hunk 6]
```diff
@@ -101,12 +106,14 @@ func TestTxClient_SignAndBroadcast_Succeeds(t *testing.T) {
...
 	require.NoError(t, err)
 
 	// Publish the transaction event bytes to the events query client so that the transaction client
 	// registers the transactions as committed (i.e. removes it from the timeout pool).
+	eventsBzPublishChMu.Lock()
 	eventsBzPublishCh <- either.Success[[]byte](txEventBz)
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:115]:
```suggestion
	eventsBzPublishCh <- either.Success[[]byte](txResultBz)
```

File: pkg/client/tx/client_test.go [hunk 7]
```diff
@@ -43,6 +45,9 @@ func TestTxClient_SignAndBroadcast_Succeeds(t *testing.T) {
 		// testtx.NewOneTimeTxTxContext helper function. The same reference needs
 		// to be used across the expectations that are set on the transactions context mock.
 		expectedTx cometbytes.HexBytes
+		// eventsBzPublishChMu is a mutex that protects eventsBzPublishCh from concurrent access
+		// as it is expected to be updated in a mock method but is also sent on in the test.
+		eventsBzPublishChMu = new(sync.Mutex)
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:50]:
```suggestion
		// txResultsBzPublishChMu is a mutex that protects txResultsBzPublishCh from concurrent access
		// as it is expected to be updated in a mock method but is also sent on in the test.
		txResultsBzPublishChMu = new(sync.Mutex)
```

File: pkg/client/tx/client_test.go [hunk 8]
```diff
@@ -54,7 +59,7 @@ func TestTxClient_SignAndBroadcast_Succeeds(t *testing.T) {
 	keyring, signingKey := testkeyring.NewTestKeyringWithKey(t, testSigningKeyName)
 
 	eventsQueryClient := testeventsquery.NewOneTimeTxEventsQueryClient(
-		ctx, t, signingKey, &eventsBzPublishCh,
+		ctx, t, signingKey, eventsBzPublishChMu, &eventsBzPublishCh,
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:62]:
```suggestion
		ctx, t, signingKey, txResultsBzPublishChMu, &txResultsBzPublishCh,
```

File: pkg/client/tx/client_test.go [hunk 9]
```diff
@@ -43,6 +45,9 @@ func TestTxClient_SignAndBroadcast_Succeeds(t *testing.T) {
 		// testtx.NewOneTimeTxTxContext helper function. The same reference needs
 		// to be used across the expectations that are set on the transactions context mock.
 		expectedTx cometbytes.HexBytes
+		// eventsBzPublishChMu is a mutex that protects eventsBzPublishCh from concurrent access
+		// as it is expected to be updated in a mock method but is also sent on in the test.
+		eventsBzPublishChMu = new(sync.Mutex)
 		// eventsBzPublishCh is the channel that the mock events query client
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:51]:
```suggestion
		// txResultsBzPublishCh is the channel that the mock events query client
```

File: pkg/client/tx/client_test.go [hunk 10]
```diff
@@ -101,12 +106,14 @@ func TestTxClient_SignAndBroadcast_Succeeds(t *testing.T) {
 	require.NoError(t, err)
 
 	// Construct the expected transaction event bytes from the expected transaction bytes.
-	txEventBz, err := json.Marshal(&tx.TxEvent{Tx: expectedTx})
+	txEventBz, err := json.Marshal(abci.TxResult{Tx: expectedTx})
 	require.NoError(t, err)
 
 	// Publish the transaction event bytes to the events query client so that the transaction client
 	// registers the transactions as committed (i.e. removes it from the timeout pool).
+	eventsBzPublishChMu.Lock()
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:114]:
```suggestion
	txResultsBzPublishChMu.Lock()
```

File: pkg/client/tx/client_test.go [hunk 11]
```diff
@@ -101,12 +106,14 @@ func TestTxClient_SignAndBroadcast_Succeeds(t *testing.T) {
...
 	eventsBzPublishCh <- either.Success[[]byte](txEventBz)
+	eventsBzPublishChMu.Unlock()
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:116]:
```suggestion
	txResultsBzPublishChMu.Unlock()
```

File: pkg/client/tx/client_test.go [hunk 12]
```diff
@@ -184,6 +191,9 @@ func TestTxClient_NewTxClient_Error(t *testing.T) {
 
 func TestTxClient_SignAndBroadcast_SyncError(t *testing.T) {
 	var (
+		// eventsBzPublishChMu is a mutex that protects eventsBzPublishCh from concurrent access
+		// as it is expected to be updated in a mock method but is also sent on in the test.
+		eventsBzPublishChMu = new(sync.Mutex)
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:196]:
```suggestion
		// txResultsBzPublishChMu is a mutex that protects txResultsBzPublishCh from concurrent access
		// as it is expected to be updated in a mock method but is also sent on in the test.
		txResultsBzPublishChMu = new(sync.Mutex)
```

File: pkg/client/tx/client_test.go [hunk 13]
```diff
@@ -184,6 +191,9 @@ func TestTxClient_NewTxClient_Error(t *testing.T) {
...
 		// eventsBzPublishCh is the channel that the mock events query client
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:197]: (this one too)

File: pkg/client/tx/client_test.go [hunk 14]
```diff
@@ -203,7 +213,7 @@ func TestTxClient_SignAndBroadcast_SyncError(t *testing.T) {
 	// NewTxClient call to fail, we don't need to set any expectations
 	// on this mock.
 	eventsQueryClient := testeventsquery.NewOneTimeTxEventsQueryClient(
-		ctx, t, signingKey, &eventsBzPublishCh,
+		ctx, t, signingKey, eventsBzPublishChMu, &eventsBzPublishCh,
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:216]:
```suggestion
		ctx, t, signingKey, txResultsBzPublishChMu, &txResultsBzPublishCh,
```

File: pkg/client/tx/client_test.go [hunk 15]
```diff
@@ -254,6 +264,9 @@ func TestTxClient_SignAndBroadcast_CheckTxError(t *testing.T) {
 		// by the transaction client. It is computed and assigned in the
 		// testtx.NewOneTimeErrCheckTxTxContext helper function.
 		expectedErrMsg string
+		// eventsBzPublishChMu is a mutex that protects eventsBzPublishCh from concurrent access
+		// as it is expected to be updated in a mock method but is also sent on in the test.
+		eventsBzPublishChMu = new(sync.Mutex)
 		// eventsBzPublishCh is the channel that the mock events query client
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:270]:
```suggestion
		// txResultsBzPublishChMu is a mutex that protects txResultsBzPublishCh from concurrent access
		// as it is expected to be updated in a mock method but is also sent on in the test.
		txResultsBzPublishChMu = new(sync.Mutex)
		// txResultsBzPublishCh is the channel that the mock events query client
```

File: pkg/client/tx/client_test.go [hunk 16]
```diff
@@ -254,6 +264,9 @@ func TestTxClient_SignAndBroadcast_CheckTxError(t *testing.T) {
...
 		// will use to publish the transactions event bytes. It is used near the end of
 		// the test to mock the network signaling that the transactions was committed.
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:272]: (another one)

File: pkg/client/tx/client_test.go [hunk 17]
```diff
@@ -265,7 +278,7 @@ func TestTxClient_SignAndBroadcast_CheckTxError(t *testing.T) {
 	keyring, signingKey := testkeyring.NewTestKeyringWithKey(t, testSigningKeyName)
 
 	eventsQueryClient := testeventsquery.NewOneTimeTxEventsQueryClient(
-		ctx, t, signingKey, &eventsBzPublishCh,
+		ctx, t, signingKey, eventsBzPublishChMu, &eventsBzPublishCh,
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:281]:
```suggestion
		ctx, t, signingKey, txResultsBzPublishChMu, &txResultsBzPublishCh,
```

File: pkg/client/tx/client_test.go [hunk 18]
```diff
@@ -317,6 +330,9 @@ func TestTxClient_SignAndBroadcast_Timeout(t *testing.T) {
 		// by the transaction client. It is computed and assigned in the
 		// testtx.NewOneTimeErrCheckTxTxContext helper function.
 		expectedErrMsg string
+		// eventsBzPublishChMu is a mutex that protects eventsBzPublishCh from concurrent access
+		// as it is expected to be updated in a mock method but is also sent on in the test.
+		eventsBzPublishChMu = new(sync.Mutex)
 		// eventsBzPublishCh is the channel that the mock events query client
 		// will use to publish the transaction event bytes. It is used near the end of
 		// the test to mock the network signaling that the transaction was committed.
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:338]: (again)

File: pkg/client/tx/client_test.go [hunk 19]
```diff
@@ -317,6 +330,9 @@ func TestTxClient_SignAndBroadcast_Timeout(t *testing.T) {
 		// by the transaction client. It is computed and assigned in the
 		// testtx.NewOneTimeErrCheckTxTxContext helper function.
 		expectedErrMsg string
+		// eventsBzPublishChMu is a mutex that protects eventsBzPublishCh from concurrent access
+		// as it is expected to be updated in a mock method but is also sent on in the test.
+		eventsBzPublishChMu = new(sync.Mutex)
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:335]:
```suggestion
		// txResultsBzPublishChMu is a mutex that protects txResultsBzPublishCh from concurrent access
		// as it is expected to be updated in a mock method but is also sent on in the test.
		txResultsBzPublishChMu = new(sync.Mutex)
```

File: pkg/client/tx/client_test.go [hunk 20]
```diff
@@ -328,7 +344,7 @@ func TestTxClient_SignAndBroadcast_Timeout(t *testing.T) {
 	keyring, signingKey := testkeyring.NewTestKeyringWithKey(t, testSigningKeyName)
 
 	eventsQueryClient := testeventsquery.NewOneTimeTxEventsQueryClient(
-		ctx, t, signingKey, &eventsBzPublishCh,
+		ctx, t, signingKey, eventsBzPublishChMu, &eventsBzPublishCh,
```
Comment (bryanchriswhite) [inline pkg/client/tx/client_test.go:347]:
```suggestion
		ctx, t, signingKey, txResultsBzPublishChMu, &txResultsBzPublishCh,
```

**Pull request merged**